### PR TITLE
Issue 24979 template builder implement the box component UI

### LINKED ISSUE #24979: Template Builder: Implement the Box Component
### Parent
#24535 

### User Story
As a user of the Template Builder, I want to be able to add a box component, so I can create more complex layouts.

### Acceptance Criteria
- The box component should be implemented using the design provided.
- The box component should use primeng components with it variants

### Proposed Objective
Core Features

### Proposed Priority
Priority 3 - Average

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.:
[Figma](https://www.figma.com/file/ECuBLIRWZjEzLzmlPrA0Rk/Template-Builder?type=design&node-id=344%3A16814&t=rH9rs6dgDWwkwwcf-1)

image

### Assumptions & Initiation Needs
- The design for the box component has been provided.
- The primeng components have been identified.

### Quality Assurance Notes & Workarounds
N/A

### Sub-Tasks & Estimates:
- Implement the design for the box component
- Integrate the box component with the primeng components

### PR DESCRIPTION
**Pull request by manuelrojas**
### <samp>🤖 Generated by Copilot at 74dff6f</samp>

This pull request adds a new component called `TemplateBuilderBoxComponent` that can display a list of items and some buttons in different sizes and layouts. The component is part of the template builder feature, which allows users to create and edit templates. The pull request includes the HTML template, the SCSS styles, the TypeScript class, the unit tests and the storybook story for the component, as well as the module and SCSS changes to import and use the component.

### Checklist
- [ ] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)

### Additional Info
** any additional useful context or info **

### Screenshots

image

https://github.com/dotCMS/core/assets/3438705/1c4395e1-432d-4853-ac87-6091d2fb0daf

### CODE REVIEW

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.scss [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        display: flex;
+        align-items: center;
+        justify-content: space-between;
+
+        p {
+            overflow: hidden;
+            text-overflow: ellipsis;
+            white-space: nowrap;
+        }
```
Comment (zJaaal) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.scss:39]: On top of this file you can do `@import "mixins";` and then here do `@include truncate-text;` that way we use the truncate mixin created on global variables.

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.html [hunk 2]
```diff
@@ -0,0 +1,59 @@
+<ng-template [ngIf]="size !== 'small'">
+    <p-card
+        [ngClass]="{
+            'template-builder-box__medium': size === 'medium',
```
Comment (zJaaal) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.html:4]: In BEM standard, the variations are represented using `--` so in this case should be `template-builder-box--medium`

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+@Component({
+    selector: 'dotcms-host-component',
+    template: ` <dotcms-template-builder-box [size]="size"> </dotcms-template-builder-box>`
+})
+class TemplateBuilderBoxHostComponent {
+    size = 'large';
+}
+
+describe('TemplateBuilderBoxComponent', () => {
```
Comment (zJaaal) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts:22]: We should use jest as we discussed, you need to import `describe, it, expect` from `@jest/globals`

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    it('should create the component', () => {
+        expect(component).toBeTruthy();
+    });
+
+    it('should render with default size', () => {
+        expect(de.nativeElement.querySelector('.template-builder-box__large')).toBeTruthy();
+    });
+
+    it('should render with medium size and update the class', async () => {
```
Comment (zJaaal) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts:54]: do we need this async?

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        component.size = TemplateBuilderBoxSize.medium;
+        hostFixture.detectChanges();
+
+        expect(de.nativeElement.querySelector('.template-builder-box__medium')).toBeTruthy();
+    });
+
+    it('should render with small size and update the class', async () => {
```
Comment (zJaaal) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts:61]: do we need this async?

**Review by zJaaal — COMMENTED** [review]
See my comments

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.html [hunk 6]
```diff
@@ -0,0 +1,59 @@
+<ng-template [ngIf]="size !== 'small'">
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.html:1]: There is ngif/else for example:

```
<div *ngIf="isLoggedIn; else loggedOut">
  Welcome back, friend.
</div>

<ng-template #loggedOut>
  Please friend, login.
</ng-template>
```
More info: https://ultimatecourses.com/blog/angular-ngif-else-then

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.html [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    <p-card
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.html:2]: Not sure about using card here, any reason I am missing?

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts (hunk 3 above)
Comment (fmontes) [reply core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts:22]: If this is running it means our jest config allow us to use this without explicit imports.

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        de = hostFixture.debugElement;
+        hostFixture.detectChanges();
+    });
+
+    it('should create the component', () => {
+        expect(component).toBeTruthy();
+    });
+
+    it('should render with default size', () => {
+        expect(de.nativeElement.querySelector('.template-builder-box__large')).toBeTruthy();
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts:51]: We didn't tell you but we don't use styles selector for tests we do `data-testId="xxxx"` and then use that as a selector.

Also, Angular have a class to select elements: https://angular.io/api/platform-browser/By

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts (hunk 4 above)
Comment (fmontes) [reply core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts:54]: We might want to use Spectator here: https://ngneat.github.io/spectator/

@oidacra and @hmoreras are already using it in the Experiment portlet and one of the cool thing is that you can create host parent components.

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.stories.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+const Template: Story<TemplateBuilderBoxComponent> = (args: TemplateBuilderBoxComponent) => ({
+    props: args
+});
+
+export const Primary = Template.bind({});
+
+Primary.args = {
+    size: 'large'
+};
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.stories.ts:28]:
```suggestion
export const Small = Template.bind({});
export const Medium = Template.bind({});
export const Large = Template.bind({});

Small.args = {
    size: 'small'
};

Medium.args = {
    size: 'medium'
};

Large.args = {
    size: 'large'
};

```

This way we can have all three in this story.

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+})
+export class TemplateBuilderBoxComponent {
+    items = [
+        { label: 'demo.dotcms.com' },
+        { label: 'System Container' },
+        { label: 'demo.dotcms.com' },
+        { label: 'demo.dotcms.com' },
+        { label: 'demo.dotcms.com' },
+        { label: 'demo.dotcms.com' }
+    ];
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts:30]: We need to discuss if this comes from an input or from where. But whatever we do we can mock it in Storybook.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts (hunk 4 above)
Comment (oidacra) [reply core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.spec.ts:54]: this file has an example `dot-experiment-options.component.spec.ts`

File: core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts [hunk 11]
```diff
@@ -20,14 +20,8 @@ export enum TemplateBuilderBoxSize {
...
 export class TemplateBuilderBoxComponent {
-    items = [
-        { label: 'demo.dotcms.com' },
-        { label: 'System Container' },
-        { label: 'demo.dotcms.com' },
-        { label: 'demo.dotcms.com' },
-        { label: 'demo.dotcms.com' },
-        { label: 'demo.dotcms.com' }
-    ];
+    @Input() items = [];
```
Comment (zJaaal) [inline core-web/libs/template-builder/src/lib/components/template-builder/components/template-builder-box/template-builder-box.component.ts:22]: You can use the type `DotContainers` from `models/models.ts`

**Review by zJaaal — COMMENTED** [review]
.

**Review by zJaaal — APPROVED** [review]

**Review by oidacra — APPROVED** [review]

**Review by fmontes — APPROVED** [review]

**Pull request merged**